### ISSUE TITLE
ICE fails to detect snap installed browsers

### ISSUE THREAD

**Issue by zeus10000** [peppermintos/ice#47]
Hello,
I have been using ICE with browsers which were installed as deb packages. Recently I moved to use browsers which are installed from snap packages. ICE does not detect any browsers installed in the system. I tried both Firefox and Chromium. Is this something ICE supports?

**Comment by AndydeCleyre** [comment]
What are the absolute paths to the browser executables?

**Comment by zeus10000** [comment]
/snap/bin/firefox
/snap/bin/chromium

**Comment by AndydeCleyre** [comment]
IMO this issue should be rolled into a generalized one for better browser detection. @zeus10000 can we assume `/snap/bin` is in the `PATH`?

**Comment by zeus10000** [comment]
I believe yes, since all the snap installed executables are place under /snap/bin. I changed the path for firefox `_FIREFOX_BIN = "/snap/bin/firefox"` but unable to open the application and it throws error.
```Firefox is already running, but is not responding. To open a new window, you must first close the existing Firefox process, or restart your system```

[Desktop Entry]
Version=1.0
Name=Office 365
Comment=Office 365 (Ice SSB)
Exec=firefox --class ICE-SSB-office --profile /home/xxx/.local/share/ice/firefox/office --no-remote https://www.office.com/
IceFirefox=office
Terminal=false
X-MultipleArgs=false
Type=Application
Icon=/home/xxx/.local/share/ice/office.ico
Categories=GTK;Network;
MimeType=text/html;text/xml;application/xhtml_xml;
StartupWMClass=ICE-SSB-office
StartupNotify=true

**Comment by AndydeCleyre** [comment]
Sounds like you still have a firefox process (with that office profile, I think) still running. I don't know if the snap-ness is introducing weird limitations, though. Can you provide usual arguments to it, like

```bash
/snap/bin/firefox -ProfileManager
```
?

**Comment by zeus10000** [comment]
Yes, I can. Looks like the issue might be snap applications cannot access `.local` directory to create profile. I created a profile in default firefox folder and it worked.
```[Desktop Entry]
Version=1.0
Name=Office 365
Comment=Office 365 (Ice SSB)
Exec=firefox --class ICE-SSB-office --profile /home/xxx/snap/firefox/common/.mozilla/firefox/n2gs7f64.office --no-remote https://www.office.com/
IceFirefox=office
Terminal=false
X-MultipleArgs=false
Type=Application
Icon=/home/xxx/.local/share/ice/office.ico
Categories=GTK;Network;
MimeType=text/html;text/xml;application/xhtml_xml;
StartupWMClass=ICE-SSB-office
StartupNotify=true```

**Comment by rhein** [comment]
Snaps runs completely encapsulated in their own world, it is quite tricky to do anything inside that from the outside (I had several problems with nextcloud as snap and finally ditched it for the normal version).
I guess it could be fixed by creating the profile in the snap location if the binary is detected in /bin/snap/ but I am not sure if this will work out with file permissions for the ice app.
What are the ownership and permissions of /home/xxx/snap directory, is that xxx:snap 775?

**Comment by zeus10000** [comment]
no, it is under user group and not under snap group `xxx:xxx` `755`. On the other hand ICE app is owned by `root:root` `755`. I believe if ICE can create profiles under snap directory it should work without issues.

**Comment by AndydeCleyre** [comment]
Since this issue does indeed involve snap-specific difficulties, I'm moving the generalized request for better bin detection to #48

**Comment by rhein** [comment]
Alright, I will check this and try a solution in my fork (just did some changes for easier support of additional browsers after adding brave anyway). Should be an easy fix really.

**Comment by zeus10000** [comment]
Thank You!

**Comment by rhein** [comment]
I don't quite get why the snap browser would not be able to use directories under /home/xxx/.local/... (where the ice profile is) but can access stuff with seemingly the same permissions under /home/xxx/snap/...
Will investigate this further, as it would be much cleaner if we don't need to have profiles scattered all around the place.

**Comment by Coeur-Noir** [comment]
Snap app don't use standard location for their settings and config files.
So nothing regarding snap under ~/.config or ~/.cache or ~/.local or ~/.app_name

They will use /home/snap/app/…

For chromium it's something like
/home/django/$USER/chromium/common/.cache

**Comment by Coeur-Noir** [comment]
And still works like a charm in Ubuntu Budgie 20.04 - with Chromium snap and specific profile ;-)

**Comment by PizzaLovingNerd** [comment]
It worked fine for me with Chromium when I ran Ubuntu. But Firefox's Flatpak doesn't work

**Comment by Coeur-Noir** [comment]
You may check where flatpak stores things, regarding Chromium.

**Comment by pshem** [comment]
The chromium snap is installed in classic mode (unconfined) by default. Firefox by default is confined. If you install the firefox snap with the --classic flag, it will be able to access all files on your system. See https://snapcraft.io/docs/snap-confinement for a detailed overview

**Comment by Coeur-Noir** [comment]
> The chromium snap is installed in classic mode (unconfined) by default. Firefox by default is confined...

Oh no, neither snap Chromium nor Firefox won't be installed in classic mode. They both are confined by default and shouldn't allow --classic installation.

Look for interfaces in snap manual : removable-media, home and desktop. This is the proper way to connect a snap to other places.
